pkg/ccl/changefeedccl/sink_kafka_v2.go

```diff
@@ -0,0 +1,614 @@
+package changefeedccl
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"hash/fnv"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/IBM/sarama"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/admission"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
+	"github.com/rcrowley/go-metrics"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/pkg/kversion"
+	"github.com/twmb/franz-go/pkg/sasl"
+	sasloauth "github.com/twmb/franz-go/pkg/sasl/oauth"
+	saslplain "github.com/twmb/franz-go/pkg/sasl/plain"
+	saslscram "github.com/twmb/franz-go/pkg/sasl/scram"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+// newKafkaSinkClientV2 creates a new kafka sink client. It is a thin wrapper
+// around the kgo client for use by the batching sink. It's not meant to be
+// invoked on its own, but rather through makeKafkaSinkV2.
+func newKafkaSinkClientV2(
+	ctx context.Context,
+	clientOpts []kgo.Opt,
+	batchCfg sinkBatchConfig,
+	bootstrapAddrs string,
+	settings *cluster.Settings,
+	knobs kafkaSinkV2Knobs,
+	mb metricsRecorderBuilder,
+) (*kafkaSinkClientV2, error) {
+
+	baseOpts := []kgo.Opt{
```

Comment:
a kind of open question here is whether or not to allow idempotency. like sarama, kgo will enable idempotency on its own if its other options permit it. idempotency allows for a few benefits:
- increase num inflights requests from 1 -> 5 without endangering ordering guarantees
- reduce duplicate messages in some circumstances

but it has a few gotchas
- there are some error conditions which, if encountered, kgo will retry / block forever, because not doing so would result in ordering violations. what are these? sounds complicated
- permission to do idempotent writes can be disabled at the server level and it's unclear to me how kgo would handle this if it encountered it.

it would be simpler overall to disable it, but i dont think we really have to. curious what you all think. see:
- https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#DisableIdempotentWrite
- https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#RecordRetries
- https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#RecordDeliveryTimeout
- https://www.conduktor.io/kafka/idempotent-kafka-producer/

i'm in favour of leaving it on, myself.